Leanplum-SDK/Classes/Leanplum.m

```diff
@@ -3620,14 +3591,68 @@ - (void)trackMessageEvent:(NSString *)event withValue:(double)value andInfo:(NSS
             andParameters:(NSDictionary *)params
 {
     if (!IS_NOOP && _messageId) {
-        event = [self eventWithParentEventNamesFromEvent:event];
-        if (event) {
-            [Leanplum track:event
+        LP_TRY
+        NSMutableString *fullEventName = [NSMutableString string];
+        LPActionContext *context = self;
+        NSMutableArray *parents = [NSMutableArray array];
+        while (context->_parentContext != nil) {
+            [parents addObject:context];
+            context = context->_parentContext;
+        }
+        NSString *actionName;
+        NSMutableDictionary *messagesForTrack= [[NSMutableDictionary alloc] init];
+        
+        for (NSInteger i = parents.count - 1; i >= -1; i--) {
+            NSString *messageIdForTrack = _messageId;
+            context=self;
+            if (i > -1) {
+                context=(LPActionContext *) parents[i];
+                actionName =  context->_key;
+                if(context->_chainedToExisting){
+                    messageIdForTrack = context->_parentContext->_messageId;
+                }
+            } else {
+                actionName = event;
+            }
+            if (actionName == nil) {
+                fullEventName = nil;
+                break;
+            }
+            actionName = [actionName stringByReplacingOccurrencesOfString:@" action"
+                                                               withString:@""
+                                                                  options:NSCaseInsensitiveSearch
+                                                                    range:NSMakeRange(0,
+                                                                                      actionName.length)
+                          ];
+            if (fullEventName.length) {
+                [fullEventName appendString:@" "];
+            }
+            [fullEventName appendString:actionName];
+            for(NSString *messageKey in messagesForTrack.allKeys){
```

Comment:
Let's make a new method just for this similar to  eventWithParentEventNamesFromEvent:
Something like
- (NSDictionary *)messagesToTrackFromEvent:(NSString *)event;
Have a clear documentation on when this method will be used: Chained message for Accept Accept and such.

Have another method
- (void)enumerateParentContextWithBlock:(void (^)(LPActionContext *context, NSInteger index, NSString *actionName));
That will be used both by messagesToTrackFromEvent: and messagesToTrackFromEvent: